clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -19,46 +19,85 @@
+ * This is responsible for:
+ * <li>Keeping member info (ex. member id, member epoch, assignment, etc.)</li>
+ * <li>Keeping member state as defined in {@link MemberState}.</li>
+ * <p/>
+ * Member info and state are updated based on the heartbeat responses the member receives.
  */
 public class MembershipManagerImpl implements MembershipManager {
 
+    /**
+     * Group ID of the consumer group the member will be part of, provided when creating the current
+     * membership manager.
+     */
     private final String groupId;
+
+    /**
+     * Group instance ID to be used by the member, provided when creating the current membership manager.
+     */
     private final Optional<String> groupInstanceId;
+
+    /**
+     * Member ID assigned by the server to the member, received in a heartbeat response when
+     * joining the group specified in {@link #groupId}
+     */
     private String memberId;
+
+    /**
+     * Current epoch of the member. It will be set to 0 by the member, and provided to the server
+     * on the heartbeat request, to join the group. It will be then maintained by the server,
+     * incremented as the member reconciles and acknowledges the assignments it receives. It will
+     * be reset to 0 if the member gets fenced.
+     */
     private int memberEpoch;
+
+    /**
+     * Current state of this member as part of the consumer group, as defined in {@link MemberState}
+     */
     private MemberState state;
+
+    /**
+     * Assignor selection configured for the member, that will be sent out to the server on the
+     * {@link ConsumerGroupHeartbeatRequest}. This will default to using server-side assignor,
+     * letting the server choose the specific assignor implementation to use.
+     */
     private AssignorSelection assignorSelection;
 
     /**
      * Assignment that the member received from the server and successfully processed.
      */
     private ConsumerGroupHeartbeatResponseData.Assignment currentAssignment;
+
     /**
      * Assignment that the member received from the server but hasn't completely processed
      * yet.
      */
     private Optional<ConsumerGroupHeartbeatResponseData.Assignment> targetAssignment;
+
     /**
-     * Latest assignment that the member received from the server while a {@link #targetAssignment}
-     * was in process.
+     * Logger.
      */
-    private Optional<ConsumerGroupHeartbeatResponseData.Assignment> nextTargetAssignment;
+    private final Logger log;
 
-    public MembershipManagerImpl(String groupId) {
-        this(groupId, null, null);
+    public MembershipManagerImpl(String groupId, LogContext logContext) {
+        this(groupId, null, null, logContext);
     }
 
-    public MembershipManagerImpl(String groupId, String groupInstanceId, AssignorSelection assignorSelection) {
+    public MembershipManagerImpl(String groupId,
+                                 String groupInstanceId,
+                                 AssignorSelection assignorSelection,
```

Comment:
Interesting simplification, makes sense to me. So it's fair to say that from the consumer point of view, there could be really only one Optional<String> `clientAssignor`. That will have a class name of the client assignor to use or empty (empty meaning that a server-side assignor will be used, and the server will choose which one), correct? Sounds good to me.

Reply:
Just to double check, this could also include the name of a server-side assignor to use, specified by the consumer, right? (to be provided in the HB request on the `server_assignor` field).

Reply:
We will only support server-side assignors for now. The config that we need to add is a nullable string. When null, the server selects the assignor. Otherwise, we pass the name in the HB request as you said.